### PR TITLE
Add Shelley keypair generation command to cardano-cli

### PR DESCRIPTION
**Pull request by intricate**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Types.hs [hunk 1]
```diff
@@ -48,9 +49,19 @@ data KeyPair
...
+  | RegularShelleyKey
+  deriving (Generic, NFData, Show)
+  deriving anyclass NoUnexpectedThunks
+
+-- | Convert a 'ShelleyKeyDiscriminator' to a 'KeyDiscriminator'.
+toShelleyLedgerKeyDiscriminator :: ShelleyKeyDiscriminator -> KeyDiscriminator
+toShelleyLedgerKeyDiscriminator skd =
+  case skd of
+    GenesisShelleyKey -> Genesis
+    RegularShelleyKey -> Regular
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Types.hs:64]: Do we want conversion functions in the `Types` module or should they be placed elsewhere?

File: cardano-cli/src/Cardano/CLI/Ops.hs [hunk 2]
```diff
@@ -270,6 +271,7 @@ data CliError
   | UpdateProposalSubmissionError !RealPBFTError
   | VerificationKeyDeserialisationFailed !FilePath !Text
   | FileNotFoundError !FilePath
+  | CardanoApiError !ApiError
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Ops.hs:274]: @erikd: Should I define more specific errors? e.g. `ShelleyKeyPairGenerationError`.

File: cardano-api/src/Cardano/Api/Types.hs (hunk 1 above)
Comment (erikd) [reply cardano-api/src/Cardano/Api/Types.hs:64]: Its fine where it is.

File: cardano-cli/src/Cardano/CLI/Ops.hs (hunk 2 above)
Comment (erikd) [reply cardano-cli/src/Cardano/CLI/Ops.hs:274]: Probably only need one constructor for `ApiError`.

File: cardano-api/src/Cardano/Api/Types.hs (hunk 1 above)
Comment (intricate) [reply cardano-api/src/Cardano/Api/Types.hs:64]: This is unused, my bad. Dropping it from the changeset.

File: cardano-cli/src/Cardano/CLI/Shelley/Run.hs [hunk 3]
```diff
@@ -26,5 +29,6 @@ runShelleyClientCommand cc =
 
 
 runShelleyKeyGenerate :: OutputFile -> ByteString -> ExceptT CliError IO ()
-runShelleyKeyGenerate fpath title =
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run.hs:29]: `_title` should be removed. Happy to green light aftet this change.

File: cardano-cli/src/Cardano/CLI/Shelley/Run.hs [hunk 4]
```diff
@@ -26,5 +29,6 @@ runShelleyClientCommand cc =
...
-  liftIO . putStrLn $ "runShelleyKeyGenerate (" ++ show fpath ++ ") " ++ show title
+runShelleyKeyGenerate (OutputFile fpath) _title = do
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run.hs:32]: `_title` should be removed. Happy to green light after this change.

File: cardano-cli/src/Cardano/CLI/Shelley/Run.hs [hunk 5]
```diff
@@ -26,5 +29,6 @@ runShelleyClientCommand cc =
...
+  kp <- liftIO $ shelleyGenKeyPair RegularShelleyKey
+  firstExceptT CardanoApiError (ExceptT (writeKeyPair fpath kp))
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run.hs:34]: You can use `liftIO` or `newExceptT` here

Comment (intricate) [reply cardano-cli/src/Cardano/CLI/Shelley/Run.hs:34]: I don't mind switching it to `newExceptT`, but:

```
-- | Constructor for computations in the ExceptT monad.
-- (The inverse of 'runExceptT').
newExceptT :: m (Either x a) -> ExceptT x m a
newExceptT =
  ExceptT
```

Comment (intricate) [reply cardano-cli/src/Cardano/CLI/Shelley/Run.hs]: Switched it to `newExceptT`.

**Comment by dcoutts** [comment]
bors merge

**Pull request merged**